Introduction to Go

craaaa opened the pull request:
Added a section that introduces Go; view the rendered version [here](https://github.com/craaaa/learningresources/blob/intro-to-go/contents/go/Go.md)!

@darrenwee -- since you did your lightning talk on Go, would appreciate if you could take a look!

craaaa committed:
Add Intro to Go section

craaaa committed:
Add author to Go.md

craaaa committed:
Add link to Go.md in ToC.md

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+Goroutines can also be started for anonymous functions:
+```go
+go func(msg string) {
+  fmt.Println(msg)
````
Go uses hard tabs for indentation, not spaces.

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+In the example below, the `Rectangle` type implements the interface `TwoDimensional` by implementing the methods `area()` and `perim()` that are specified in the interface.
+
+```go
+type TwoDimensional interface {
````
Run `go fmt` on this code snippet? Doesn't look like the canonical style to me.

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+
+If you are interested in delving deeper into using Go's concurrency features extensively, Google developers have put out video presentations on Go's [basic](https://www.youtube.com/watch?v=f6kdp27TYZs) and [advanced concurrency patterns](https://www.youtube.com/watch?v=QDDwwePbDtw). This [code walkthrough](https://golang.org/doc/codewalk/sharemem/) provides an annotated example of how Go's memory ...
+
+### Effective Go
```
Rename this section to something more relevant to formatting, like `### Canonical Code Style`.

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+
+#### Channels
+
+Channels are used to fulfill Go's philosophy on concurrent software: "don't communicate by sharing memory; share memory by communicating".
```
Specify that Go relies primarily on the message passing pattern to explain the quote in software engineering terms?

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+}("a message")  // starts a goroutine that prints "a message"
+```
+
+A goroutine is not its own thread; instead, goroutines are dynamically multiplexed onto threads as required to keep them running. This makes them lightweight, so having a large number of goroutines is feasible. In practice, goroutines behave similarly to very cheap threads.
````
You can also specify that goroutines start out with a very small stack (on the order of kBs instead of an OS thread in MBs); see [here](https://stackoverflow.com/questions/41906357/why-are-goroutines-much-cheaper-than-threads-in-other-languages)

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+
+```go
+func run() {
+  \\ does something
````
A comment in Go uses `//` instead of `\\`

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+More information on these three mechanisms and their uses can be found on the [Go blog](https://blog.golang.org/defer-panic-and-recover).
+
+### Interfaces
+Although Go has types and methods and allows pseudo-object-oriented style of programming, type hierarchy does not exist in Go. Instead, Go uses interfaces to specify methods that types should implement, grouping types by functionality rather than using notions of inheritance. Types implement interfaces by implementing the methods in the interface, and do not need to explicitly specify which inter ...
```
Use the magic phrase `composition over inheritance` since it's a pattern.

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+To create dynamically-sized arrays, Go introduces the concept of a slice. Slices are not arrays; rather, they are data structures that describe a piece of a separately-stored array. Slices are the topic of several Go blog entries; you can read more about [the internal layout of slices](https://blog.golang.org/go-slices-usage-and-internals) and [the mechanics of using slices](https://blog.golang.o ...
+
+### `defer`, `panic` and `recover`
+In addition to traditional control flow mechanisms such as `if`, `for` and `switch`, Go also provides three other ways to alter the flow of a program: `defer`, `panic`, `recover`. These mechanisms can be conceived in terms of exception handling in Java and C++, but also have more general purpose uses.
```
I think this section needs to be a lot more precise and rescoped.

Exception handling is kind of special in Go because you rely on multiple returns, e.g. some function returns `(int, int, String, err)` and we always check `if err != nil` to perform exception handling. `panic` is only used if `err != nil` _and_ we don't know what to do. Tl;dr `panic` and `recover` only describe a subset of exception handling
https://blog.golang.org/error-handling-and-go

You could talk about `defer` here, and move `panic` and `recover` into a separate section dedicated to exception handling. I think it's worth doing so because exception handling is quite different in Go compared to other languages.

You can also include an example of `defer` since it's not a common language feature (I think).

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+func (r Rectangle) perim() {
+  return r.width * 2 + r.height * 2
+}
+```
````
Add an example, like `type Circle struct` which implements `area()` but not `perim()` to show a negative example.

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+
+If you're unconvinced about Go, [the Go playground](https://play.golang.org/) can be used as a sandbox in which to write, build and execute code without installing Go on your machine.
+
+## Using Go
```
Add a section on variable declaration, e.g. the difference between using `var`, `:=` for inference, etc. You can also cover the reversed adjective order, e.g. in Java we specify `String message` but in Go we say `message string`. https://blog.golang.org/gos-declaration-syntax

You can also cover pointers in the same section, since they're a thing in Go but their declaration syntax is also different.

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+    value int // its value
+}
+```
+Running `$ gofmt` in the same directory as the source file will line up comments and correct the source code's indentation:
````
I think if you're specifying a path (like current directory) you need to use `$ go fmt` instead. Can you double-check?

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+
+More information on these three mechanisms and their uses can be found on the [Go blog](https://blog.golang.org/defer-panic-and-recover).
+
+### Interfaces
```
Since we're touching the OOP parts of Go, you can also add a section on how encapsulation works. https://code.tutsplus.com/tutorials/lets-go-object-oriented-programming-in-golang--cms-26540

craaaa replied:
Hm, I'm not sure I want to explicitly mention this, as the section isn't about OOP in general. There's quite a lot to be said about Go's approach to OOP, but I don't think that I want to cover that here as it's somewhat prickly. Moreover, I intended for this section to walk through potential painpoints for a new learner of Go to look out for, and I think Go's encapsulation method is pretty straightforward. Will add it to the list of links though!

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,143 @@
+```
+One useful way to think about sending and receiving data with the `<-` operator is that the data moves in the direction of the arrow.
+
+If you are interested in delving deeper into using Go's concurrency features extensively, Google developers have put out video presentations on Go's [basic](https://www.youtube.com/watch?v=f6kdp27TYZs) and [advanced concurrency patterns](https://www.youtube.com/watch?v=QDDwwePbDtw). This [code walkthrough](https://golang.org/doc/codewalk/sharemem/) provides an annotated example of how Go's memory ...
````
Include a minimum working example incorporating both goroutines and channels, e.g. asynchronously sum halves of an array.

darrenwee commented on contents/go/Go.md:
```diff
@@ -0,0 +1,143 @@
+## Using Go
+As it builds on the foundations set by many popular and widely-used languages such as C, C++, Java and Python, much of Go's syntax draws from existing implementations and will be familiar to programmers looking to learn an additional language. However, Go also diverges explicitly from these other languages; listed below are some aspects of Go that may be unfamiliar to learners.
+
+### Arrays and slices
```
`Slices`

darrenwee reviewed (changes requested):
Looks pretty good. You'll need to run `go fmt` on all the code snippets, and I've included some comments on possible things you may want to cover because I think they're quite fundamental differences between Go and most other mainstream languages.

craaaa committed:
Run go fmt on code, fix capitalization, add negative interface example

craaaa committed:
Add section on variable declaration

craaaa committed:
Update inheritance section

craaaa committed:
Update concurrency section

craaaa committed:
Edit phrasing of channel blocking

craaaa committed:
Add spaces to gofmt

craaaa committed:
Split defer, panic, recover section into two sections

craaaa commented:
Hey @darrenwee, thanks for your comments! I've made most of the changes you suggested; do check out the section and let me know what you think!

darrenwee commented on contents/go/Go.md:
````diff
@@ -0,0 +1,248 @@
+Goroutines send and receive messages through a channel using the `<-` operator.
+
+```go
+	ch <- v   // Send v to channel ch.
````
Remove indent

darrenwee reviewed (approved):
Just the one comment, then good to Go :+1: :100:

craaaa committed:
Remove indentation from example

darrenwee reviewed (approved):
LGTM

craaaa commented:
@darrenwee thanks for your help!
@damithc ready for review!

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+
+Go provides its own [installation guide](https://golang.org/doc/install) and an interactive [ tour of Go](https://tour.golang.org/). These are useful and highly comprehensive resources for programmers looking to learn the syntax and style of Go. For those who prefer to read existing code examples, [Go by Example](https://gobyexample.com/) is a collection of code samples covering a wide variety of ...
+
+If you're unconvinced about Go, [the Go playground](https://play.golang.org/) can be used as a sandbox in which to write, build and execute code without installing Go on your machine.
```
Shouldn't it be either `a sandbox to write` or `a sandbox in which you can write`?

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+
+When initializers are not used, simply declaring a variable allocates storage for the variable and initializes its value to that type's zero value. The zero value is `0` for numeric types, `false` for booleans, `""` for strings, `nil` for pointers, functions, interfaces, slices, channels, and maps.
+
+As an extension of type inference, the `:=` short assignment statement can be used inside a function to as a `var` declaration with implicit type:
```
I think this sentence is not very clear for first-timer. I need to re-read it and look at the example to figure out what you mean. I think the problem is because you are using the passive voice which makes my brain wonders who is doing what. I suggest something like `In addition to omitting the type when it can be inferred, the programmer can also eliminate the keyword var in  the variable declaration by using the := short assignment statement`.

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+Go's syntax for more complex types such as pointers, arrays and structs is also somewhat idiosyncratic, and can be explored in this [Go blog article](https://blog.golang.org/gos-declaration-syntax), or with the help of [this tutorial on pointers](http://www.golang-book.com/books/intro/8) and [this tutorial on structs](https://www.golang-book.com/books/intro/9).
+
+### Arrays and Slices
+Arrays are not often seen in Go code. This is because the size of a Go array must be declared at its creation, which limits how flexible an array can be. For instance, in the 
```
Unfinished  sentence?

tran-tien-dat commented on contents/go/Go.md:
````diff
@@ -0,0 +1,248 @@
+}
+```
+
+`defer` is frequently used for clean-up actions, such as to [close files](https://gobyexample.com/defer). Deferred functions run on panicking goroutines as well, which makes them useful for recovering from `panic`.
````
So what is `panic`? Maybe mention that it will be explained later.

tran-tien-dat commented on contents/go/Go.md:
````diff
@@ -0,0 +1,248 @@
+}
+```
+
+When calling a method that may return an error, we check if the returns `err != null` and handle the resulting error.
````
What's the difference between `null` and `nil`? You mention `null` here but `nil` is used in the code.

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+
+To deal with unexpected errors, Go also provides two mechanisms: `panic` and `recover`.
+
+- `panic` stops the ordinary flow of execution, executes any deferred functions, and returns to its caller as a call to `panic`. The `panic` continues up the stack until all functions in the goroutine have returned, after which the program crashes. `panic` is used to fail fast on errors that cannot be handled gracefully.
```
This paragraph is really hard to digest. Had to re-read a few times to realize that you wants to describe a recursive mechanism. Maybe you can add some more explanation to make it easier to understand. Best is to have an example to illustrate.

Also, up till now I have assumed that `function` is the same as `goroutine` but from this sentence I suppose there is a difference. Ok I see that it is explained in Concurrency section. I would prefer that you give a definition, which could be brief, of every term the first time you use it though.

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+To deal with unexpected errors, Go also provides two mechanisms: `panic` and `recover`.
+
+- `panic` stops the ordinary flow of execution, executes any deferred functions, and returns to its caller as a call to `panic`. The `panic` continues up the stack until all functions in the goroutine have returned, after which the program crashes. `panic` is used to fail fast on errors that cannot be handled gracefully.
+- `recover` regains control of a panicking goroutine. When used inside a deferred function, a call to recover captures the value returned by `panic` and resumes normal execution.
```
Just curious. Where does execution returns? The next line after the first `panic` that trigger all the panicking, or is it somewhere else?

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+More information on error handling can be found on the [Go blog](https://blog.golang.org/error-handling-and-go).
+
+### Interfaces
+Although Go has types and methods and allows pseudo-object-oriented style of programming, type hierarchy does not exist in Go. Instead, Go uses interfaces to specify methods that types should implement, favouring composition over inheritance. Types implement interfaces by implementing the methods in the interface, and do not need to explicitly specify which interfaces are implemented.
```
I would change `and` to `but`, because it is quite surprising for a person used to OOP paradigm to see this.

tran-tien-dat commented on contents/go/Go.md:
```diff
@@ -0,0 +1,248 @@
+
+In the example below, the `Rectangle` type implements the interface `TwoDimensional` by implementing the methods `area()` and `perim()` that are specified in the interface. Thus, instances of `Rectangle` can be used as arguments to `price`. 
+
+Meanwhile, although `Circle` implements `perim()`, it does not implement `area()`. Since it does not implement all the methods in the `TwoDimensional` interface, `Circle` does not implement `TwoDimensional`. Thus, instances of `Circle` cannot be used as arguments to `price`.
```
Maybe add code below where instances of `Rectangle` and `Circle` are used and for `Circle` the compiler gives an error.

tran-tien-dat reviewed (commented):
Hi, I wanted to learn Go so I read your book chapter and left some comments. Hope that such comments from somebody without Go experience will be useful. Overall, the book chapter is very well-written. It's not just a list of syntactic constructs but focusing on highlighting interesting features of Go. Makes me wanna try it out 👍

damithc reviewed (commented):
I'm ok with the overall structure. Request a review when you have address Dat's comments.

craaaa committed:
Edit variable declaration section

craaaa committed:
Update concurrency and error handling sections

craaaa committed:
Add details to interface section

craaaa commented:
Hi @tran-tien-dat, thanks for getting some fresh eyes on the page, and for giving such useful comments! I've made some changes to the document to make new terms more explicit (by moving sections around), and to clarify ambiguities that you pointed out.

@damithc ready for review!

tran-tien-dat reviewed (approved):
LGTM 💯

craaaa committed:
Make panic flow more verbose

damithc commented on contents/go/Go.md:
````diff
@@ -0,0 +1,261 @@
+var name = "John Smith"
+```
+
+When an initializer is not used, declaring a variable allocates storage for the variable and initializes its value to that type's zero value. The zero value is `0` for numeric types, `false` for booleans, `""` for strings, `nil` for pointers, functions, interfaces, slices, channels, and maps.
````
Can do without this sentence? Feels like its something the reader may not be interested at this stage.

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,261 @@
+Arrays are not often seen in Go code. This is because the size of a Go array must be declared at its creation, which limits how flexible an array can be. For instance, in the 
+
+To create dynamically-sized arrays, Go introduces the concept of a slice. Slices are not arrays; rather, they are data structures that describe a piece of a separately-stored array. Slices are the topic of several Go blog entries; you can read more about [the internal layout of slices](https://blog.golang.org/go-slices-usage-and-internals) and [the mechanics of using slices](https://blog.golang.o ...
+
```
Give a concrete example of slices?

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,261 @@
+# Introduction to Go
+
+Author(s): [Cara Leong](https://github.com/craaaa)
```
Can add a table of contents? a ToC makes it easier to see the high-level structure.

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,261 @@
+
+Go is an [open source project](https://github.com/golang/go). Its source code may be useful reading for those interested in learning good practices, or simply to find out more about how the language was implemented.
+
+## Getting started
```
Seems like this can be moved to the end of the chapter because at this point the reader has not yet decided whether to get started with Go.

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,261 @@
+
+If you're unconvinced about Go, you can use [the Go playground](https://play.golang.org/) to write, build and execute code without installing Go on your machine.
+
+## Using Go
```
Another way to frame this section is as `Go Features`. Then describe things that are somewhat unique to Go, but also mention why each feature matters i.e., how does it make Go better?

damithc reviewed (changes requested):
Looks good. Added a few comments.

craaaa committed:
Add table of contents

craaaa committed:
Move 'Getting Started' to bottom of file

craaaa committed:
Delete line about null initializers

craaaa committed:
Add information about why people should learn Go

craaaa committed:
Remove section on slices/arrays

Although it's an interesting difference between Go and other languages,
slices are not a "selling point" of the language. As I renamed the section
to "Go Features", I removed the section on slices to better fit the new
aim of the section.

craaaa committed:
Fix link to "why learn go"

craaaa commented:
@damithc Ready for review - made changes as suggested

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,273 @@
+Go (also known as `golang`) is a compiled, statically-typed, garbage-collected language that has special memory safety and concurrent programming features. Born out of frustration with the available languages (e.g. C, C++, Java) and environments for systems programming, Go was [conceptualized by programmers at Google](https://talks.golang.org/2012/splash.article) who sought to create a single lan ...
+
+## Why Learn Go?
+Go is [not a perfect language](https://github.com/ksimka/go-is-not-good). However, here are some reasons why you might want to learn and use Go:
```
Put this at the end of this section rather than at the beginning?
>Go is [not a perfect language](https://github.com/ksimka/go-is-not-good).

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,273 @@
+## Why Learn Go?
+Go is [not a perfect language](https://github.com/ksimka/go-is-not-good). However, here are some reasons why you might want to learn and use Go:
+
+Firstly, Go is a language built for software engineers. As it was written by software engineers at Google, Go addresses and attempts to solve some of the  pain points that exist in many commonly-used languages. For instance, the language has easy dependency management, prides itself on fast builds and has many [easy to use debugging, testing, and code-vetting tools](https://medium.com/google-clou ...
```
Alternative formatting:
>1. **Go is a language built for software engineers.** As it was written by software engineers at Google, Go addresses and attempts to solve some of the  pain points that exist in many commonly-used languages. For instance, the language has easy dependency management, prides itself on fast builds and has many [easy to use debugging, testing, and code-vetting tools](https://medium.com/google-cloud/go-tooling-in-action-eca6882ff3bc). These features make Go an easy language to use for software development.
>1. ...

damithc commented on contents/go/Go.md:
```diff
@@ -0,0 +1,273 @@
+Go's syntax for more complex types such as pointers, arrays and structs is also somewhat idiosyncratic, and can be explored in this [Go blog article](https://blog.golang.org/gos-declaration-syntax), or with the help of [this tutorial on pointers](http://www.golang-book.com/books/intro/8) and [this tutorial on structs](https://www.golang-book.com/books/intro/9).
+
+### Concurrency
+One of Go's special features is its support for concurrency. To this end, Go's standard library comes with two features that allow for easy and maintainable concurrency.
```
Other language support concurrency too. Go has _better_ support?

damithc commented on contents/go/Go.md:
````diff
@@ -0,0 +1,273 @@
+}
+```
+
+`defer` is frequently used for clean-up actions, such as to [close files](https://gobyexample.com/defer). Deferred functions run on [panicking goroutines](#error-handling) as well, which makes them useful for recovering from `panic`.
````
The mention of _panicking goroutines_  hurts the self-contained quality of this article because the reader has to go to another page before continuing to read this article. Give a quick definition (or avoid altogether)?

craaaa replied:
Decided to move this section further down in the list of Go features. The hope is that users unfamiliar with Go will have read the previous section on `panic`, which will teach them about what a panicking goroutine is, before they reach this section.

damithc reviewed (approved):
Looks good. 👍  A few cosmetic nits. I'll leave you to decide whether to follow the suggestions. Let me know when ready to merge.

craaaa committed:
Make cosmetic updates

craaaa committed:
Change section titles

Use question format to make it clear what question each section is trying
to answer.
